File: apps/web/scripts/worker-health.ts
import { env } from "../src/lib/env";
import { prisma } from "../src/lib/prisma";

const source = process.argv[2];
const maxAgeSeconds = Number(process.argv[3] ?? defaultMaxAgeSeconds(source));

function defaultMaxAgeSeconds(value: string | undefined) {
  if (value === "edge-scanner") return Math.max(300, env.kalshiCacheSeconds * 4);
  if (value === "retention-cleanup") return Math.max(7200, env.retentionWorkerIntervalSeconds * 3);
  return 900;
}

async function main() {
  if (!source) throw new Error("Usage: tsx scripts/worker-health.ts <source> [maxAgeSeconds]");
  const latest = await prisma.dataRefreshLog.findFirst({
    where: { source },
    orderBy: { startedAt: "desc" }
  });
  if (!latest) throw new Error(`No heartbeat found for ${source}.`);
  if (latest.status === "running") {
    const runningAgeSeconds = (Date.now() - latest.startedAt.getTime()) / 1000;
    if (runningAgeSeconds <= maxAgeSeconds) {
      console.log(`${source} healthy: running for ${Math.round(runningAgeSeconds)}s`);
      return;
    }
    throw new Error(`${source} has been running for ${Math.round(runningAgeSeconds)}s, over max ${maxAgeSeconds}s.`);
  }
  if (!["success", "degraded"].includes(latest.status) || !latest.completedAt) {
    throw new Error(`${source} latest status is ${latest.status}.`);
  }
  const ageSeconds = (Date.now() - latest.completedAt.getTime()) / 1000;
  if (ageSeconds > maxAgeSeconds) {
    throw new Error(`${source} latest completed heartbeat is ${Math.round(ageSeconds)}s old, over max ${maxAgeSeconds}s.`);
  }
  console.log(`${source} healthy: latest ${latest.status} ${Math.round(ageSeconds)}s ago`);
}

main()
  .catch((error) => {
    console.error(error instanceof Error ? error.message : error);
    process.exitCode = 1;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: apps/weather-service/app/worker_health.py
from __future__ import annotations

from datetime import datetime, timezone

from psycopg.rows import dict_row

from .config import settings
from .db import db_connection


def main() -> None:
    max_age_seconds = max(7200, settings.weather_ingestion_interval_seconds * 4)
    with db_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cur:
            row = cur.execute(
                """
                SELECT "status", "startedAt", "completedAt", "message"
                FROM "DataRefreshLog"
                WHERE "source" = 'weather-worker'
                ORDER BY "startedAt" DESC
                LIMIT 1
                """
            ).fetchone()
    if row is None:
        raise SystemExit("No weather-worker heartbeat found.")

    now = datetime.now(timezone.utc)
    status = str(row["status"])
    if status == "running":
        age = (now - ensure_utc(row["startedAt"])).total_seconds()
        if age <= max_age_seconds:
            print(f"weather-worker healthy: running for {int(age)}s")
            return
        raise SystemExit(f"weather-worker has been running for {int(age)}s, over max {max_age_seconds}s.")

    if status not in {"success", "degraded"} or row["completedAt"] is None:
        raise SystemExit(f"weather-worker latest status is {status}: {row['message']}")

    age = (now - ensure_utc(row["completedAt"])).total_seconds()
    if age > max_age_seconds:
        raise SystemExit(f"weather-worker latest completed heartbeat is {int(age)}s old, over max {max_age_seconds}s.")
    print(f"weather-worker healthy: latest {status} {int(age)}s ago")


def ensure_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


if __name__ == "__main__":
    main()
